File: python/dsa/unionfind/union_find.py
# Union-Find (Disjoint Set)

"""Disjoint-set connectivity with union and find.

Technique: Union-find (path compression, union by rank)
Invariant: Parent pointers form forest; ranks approximate tree height.
find/union/connected: Time O(α(n)) amortized; Space O(n).
"""


class UnionFind:
    """Disjoint-set union-find for elements 0..n-1."""

    def __init__(self, n: int) -> None:
        """Creates a structure for n elements labeled 0..n-1.

        Time: O(n); Space: O(n).
        """
        self._parent = list(range(n))
        self._rank = [0] * n
        self._components = n

    def find(self, x: int) -> int:
        """Finds the representative of x with path compression.

        Time: ~O(α(n)); Space: O(1).
        """
        if self._parent[x] != x:
            self._parent[x] = self.find(self._parent[x])
        return self._parent[x]

    def union(self, a: int, b: int) -> None:
        """Unites the sets containing a and b (weighted by rank).

        Time: ~O(α(n)); Space: O(1).
        """
        ra = self.find(a)
        rb = self.find(b)
        if ra == rb:
            return
        if self._rank[ra] < self._rank[rb]:
            self._parent[ra] = rb
        elif self._rank[ra] > self._rank[rb]:
            self._parent[rb] = ra
        else:
            self._parent[rb] = ra
            self._rank[ra] += 1
        self._components -= 1

    def connected(self, a: int, b: int) -> bool:
        """Returns True if a and b are in the same set.

        Time: ~O(α(n)); Space: O(1).
        """
        return self.find(a) == self.find(b)

    def count_components(self) -> int:
        """Returns the number of disjoint sets.

        Time: O(1); Space: O(1).
        """
        return self._components

File: javascript/src/unionfind/unionFind.js
// Union-Find (Disjoint Set)

/**
 * Disjoint-set connectivity with union and find.
 * Technique: Union-find (path compression, union by rank)
 * Invariant: Parent pointers form forest; ranks approximate tree height.
 * find/union/connected: Time O(α(n)) amortized; Space O(n).
 */

class UnionFind {
  constructor(n) {
    this._parent = Array.from({ length: n }, (_, i) => i);
    this._rank = new Array(n).fill(0);
    this._components = n;
  }

  find(x) {
    if (this._parent[x] !== x) {
      this._parent[x] = this.find(this._parent[x]);
    }
    return this._parent[x];
  }

  union(a, b) {
    let ra = this.find(a);
    let rb = this.find(b);
    if (ra === rb) {
      return;
    }
    if (this._rank[ra] < this._rank[rb]) {
      this._parent[ra] = rb;
    } else if (this._rank[ra] > this._rank[rb]) {
      this._parent[rb] = ra;
    } else {
      this._parent[rb] = ra;
      this._rank[ra] += 1;
    }
    this._components -= 1;
  }

  connected(a, b) {
    return this.find(a) === this.find(b);
  }

  countComponents() {
    return this._components;
  }
}

module.exports = { UnionFind };
